File: volleycontroller/src/main/java/com/libraries/inlacou/volleycontroller/InternetCall.java
package com.libraries.inlacou.volleycontroller;

import android.content.Context;
import android.util.Log;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.Response;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import timber.log.Timber;

/**
 * Created by inlacou on 10/09/14.
 */
public class InternetCall {

	private static final String DEBUG_TAG = InternetCall.class.getName();
	private Method method;
	private String code;
	private String url;
	private Map<String, String> params;
	private Map<String, String> headers;
	private String rawBody;
	private DefaultRetryPolicy retryPolicy;
	private ArrayList<Interceptor> interceptors;
	private File file;
	private ArrayList<VolleyController.IOCallbacks> callbacks;
	private String fileKey;
	private Object cancelTag;
	private Boolean allowLocationRedirect = true;

	public InternetCall() {
		interceptors = new ArrayList<>();
		params = new HashMap<>();
		headers = new HashMap<>();
		method = Method.GET;
		rawBody = "";
		setRetryPolicy(new DefaultRetryPolicy(10000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
	}

	public String getUrl() {
		return url;
	}

	public InternetCall setUrl(String url) {
		this.url = url;
		return this;
	}

	public File getFile() {
		return file;
	}

	public InternetCall setFile(String key, File file) {
		this.fileKey = key;
		this.file = file;
		return this;
	}

	public String getRawBody() {
		return rawBody;
	}

	public InternetCall setRawBody(String rawBody) {
		this.rawBody = rawBody;
		return this;
	}

	public InternetCall setHeaders(Map<String, String> headers) {
		this.headers = headers;
		return this;
	}

	public Map<String, String> getHeaders() {
		return headers;
	}

	public InternetCall setParams(Map<String, String> params) {
		rawBody = "";
		this.params = params;
		return this;
	}

	public Map getParams(){
		return params;
	}

	public ArrayList<VolleyController.IOCallbacks> getCallbacks() {
		return callbacks;
	}

	public InternetCall addCallback(VolleyController.IOCallbacks callback) {
		if(callbacks==null){
			callbacks = new ArrayList<>();
		}
		this.callbacks.add(callback);
		return this;
	}

	public boolean isAllowLocationRedirect(){
		return allowLocationRedirect;
	}
	
	public InternetCall setAllowLocationRedirect(Boolean b){
		allowLocationRedirect = b;
		return this;
	}
	
	public String getCode() {
		return code;
	}

	public InternetCall setCode(String code) {
		this.code = code;
		return this;
	}

	public Method getMethod() {
		return method;
	}

	public InternetCall setMethod(Method method) {
		this.method = method;
		return this;
	}

	public InternetCall replaceAccessToken(String oldAccessToken, String newAccessToken) {
		if(url!=null){
			setUrl(url.replace(oldAccessToken, newAccessToken));
		}

		Iterator it;
		if(headers!=null) {
			it = headers.entrySet().iterator();
			while (it.hasNext()) {
				Map.Entry pair = (Map.Entry) it.next();
				if(pair.getValue().toString().contains(oldAccessToken)){
					headers.put(pair.getKey().toString(), pair.getValue().toString().replace(oldAccessToken, newAccessToken));
				}
			}
		}

		if(params!=null) {
			it = params.entrySet().iterator();
			while (it.hasNext()) {
				Map.Entry pair = (Map.Entry) it.next();
				if(pair.getValue().toString().contains(oldAccessToken)){
					params.put(pair.getKey().toString(), pair.getValue().toString().replace(oldAccessToken, newAccessToken));
				}
			}
		}

		if(rawBody!=null && !rawBody.isEmpty() && rawBody.contains(oldAccessToken)) {
			rawBody = rawBody.replace(oldAccessToken, newAccessToken);
		}

		return this;
	}

	public InternetCall prebuild(){
		while(params.values().remove(null));
		while(headers.values().remove(null));
		if(rawBody==null){
			rawBody = "";
		}
		for(int i=0; i<interceptors.size(); i++){
			interceptors.get(i).intercept(this);
		}
		return this;
	}
	
	public Request build(final Context context, final com.android.volley.Response.Listener<CustomResponse> listener, com.android.volley.Response.ErrorListener errorListener) {
		if(file==null) {
			Request request = new CustomRequest(this.getMethod().value(), getUrl(), errorListener) {
				@Override
				protected void deliverResponse(Object response) {
					listener.onResponse((CustomResponse) response);
				}

				@Override
				protected Response<CustomResponse> parseNetworkResponse(NetworkResponse response) {
					CustomResponse newCustomResponse = new CustomResponse(response);
					newCustomResponse.setCode(code);
					//we set here the response (the object received by deliverResponse);
					return com.android.volley.Response.success(newCustomResponse, newCustomResponse.getChacheHeaders());
				}

				@Override
				public Map<String, String> getHeaders() throws AuthFailureError {
					if (InternetCall.this.getHeaders() != null) {
						Map<String, String> headers = InternetCall.this.getHeaders();
						for (Map.Entry<String, String> entry : headers.entrySet()) {
							Log.v(DEBUG_TAG + "." + InternetCall.this.getMethod(), "header -> " + entry.getKey() + ": " + entry.getValue());
						}
						return InternetCall.this.headers;
					}
					return super.getHeaders();
				}

				@Override
				protected Map<String, String> getParams() {
					if (InternetCall.this.getParams() != null) {
						Map<String, String> headers = InternetCall.this.getParams();
						for (Map.Entry<String, String> entry : headers.entrySet()) {
							Log.v(DEBUG_TAG + "." + InternetCall.this.getMethod(), "params -> " + entry.getKey() + ": " + entry.getValue());
						}
						return InternetCall.this.getParams();
					}
					try {
						return super.getParams();
					} catch (AuthFailureError authFailureError) {
						authFailureError.printStackTrace();
					}
					return new HashMap<>();
				}

				@Override
				public byte[] getBody() throws AuthFailureError {
					String body = InternetCall.this.getRawBody();
					if (body != null && !body.isEmpty()) {
						Log.v(DEBUG_TAG + "." + InternetCall.this.getMethod(), "body -> " + body);
						return body.getBytes();
					}
					return super.getBody();
				}
			};
			request.setRetryPolicy(retryPolicy);
			if(cancelTag!=null) request.setTag(cancelTag);
			return request;
		}else{
			Request request = new VolleyMultipartRequest(this.getMethod().value(), getUrl(), errorListener) {
				@Override
				protected void deliverResponse(Object response) {
					listener.onResponse((CustomResponse) response);
				}

				@Override
				protected Response<CustomResponse> parseNetworkResponse(NetworkResponse response) {
					CustomResponse newCustomResponse = new CustomResponse(response);
					newCustomResponse.setCode(code);
					//we set here the response (the object received by deliverResponse);
					return com.android.volley.Response.success(newCustomResponse, newCustomResponse.getChacheHeaders());
				}

				@Override
				protected Map<String, String> getParams() {
					return params;
				}

				@Override
				protected Map<String, DataPart> getByteData() {
					Map<String, DataPart> params = new HashMap<>();
					// file name could found file base or direct access from real path
					// for now just get bitmap data from ImageView
					try {
						params.put(fileKey, new DataPart(file.getName()+"."+file.getFormat(), ImageUtils.getFileDataFromBitmap(context, ImageUtils.getBitmapFromPath(file.getLocation())), file.getType().toString()+"/"+file.getFormat()));
					} catch (IOException e) {
						e.printStackTrace();
					}
					return params;
				}
			};
			request.setRetryPolicy(retryPolicy);
			if(cancelTag!=null) request.setTag(cancelTag);
			return request;
		}
	}

	public InternetCall setRetryPolicy(DefaultRetryPolicy retryPolicy) {
		this.retryPolicy = retryPolicy;
		return this;
	}

	public InternetCall setInterceptors(final ArrayList<Interceptor> interceptors){
		this.interceptors = interceptors;
		return this;
	}

	public InternetCall addInterceptors(final ArrayList<Interceptor> interceptors) {
		if(this.interceptors==null){
			this.interceptors = new ArrayList<>();
		}
		if(interceptors==null){
			return this;
		}
		this.interceptors = new ArrayList<Interceptor>(){{ addAll(interceptors); addAll(InternetCall.this.interceptors); }};
		return this;
	}

	public InternetCall addInterceptor(Interceptor interceptor){
		if(this.interceptors==null){
			this.interceptors = new ArrayList<>();
		}
		if(interceptor==null){
			return this;
		}
		this.interceptors.add(interceptor);
		return this;
	}

	public InternetCall putHeader(String key, String value) {
		if(headers==null){
			headers = new HashMap<>();
		}
		headers.put(key, value);
		return this;
	}

	public InternetCall putParam(String key, String value) {
		if(value==null){
			return this;
		}
		rawBody = "";
		if(params==null){
			params = new HashMap<>();
		}
		params.put(key, value);
		return this;
	}

	public InternetCall putHeaders(Map<String, String> headers) {
		if(this.headers==null){
			this.headers = new HashMap<>();
		}
		if(headers==null){
			return this;
		}
		this.headers.putAll(headers);
		return this;
	}

	public InternetCall putParams(Map<String, String> params) {
		while(params.values().remove(null));
		rawBody = "";
		if(this.params==null){
			this.params = new HashMap<>();
		}
		this.params.putAll(params);
		return this;
	}

	public Object getCancelTag() {
		return cancelTag;
	}

	public InternetCall setCancelTag(Object tag){
		this.cancelTag = tag;
		return this;
	}


	public enum Method{
		GET, POST, PUT, DELETE;

		int value(){
			switch (this){
				case GET:
				default:
					return Request.Method.GET;
				case POST:
					return Request.Method.POST;
				case PUT:
					return Request.Method.PUT;
				case DELETE:
					return Request.Method.DELETE;
			}
		}

		@Override
		public String toString() {
			switch (this){
				case GET:
					return "GET";
				case POST:
					return "POST";
				case PUT:
					return "PUT";
				case DELETE:
					return "DELETE";
				default:
					return "(default)GET";
			}
		}
	}

	public interface Interceptor {
		/**
		 *
		 * @param internetCall
		 * @return modified internetCall
		 */
		void intercept(InternetCall internetCall);
	}
}

File: volleycontroller/src/main/java/com/libraries/inlacou/volleycontroller/InternetCall.kt
package com.libraries.inlacou.volleycontroller

import com.android.volley.*
import com.android.volley.Response.success
import com.libraries.inlacou.volleycontroller.multipart.DataPart
import com.libraries.inlacou.volleycontroller.multipart.VolleyMultipartRequest
import org.json.JSONObject
import timber.log.Timber
import java.io.IOException

/**
 * Created by inlacou on 10/09/14.
 * Last updated by inlacou on 24/10/18.
 */
class InternetCall {
	var method: Method = Method.GET
		private set
	var code: String = ""
		private set
	var url: String? = null
		private set
	var headers: MutableMap<String, String> = mutableMapOf()
		private set
	var params: MutableMap<String, String> = mutableMapOf()
		private set
	var rawBody: String = ""
		private set
	var fileKey: String? = null
		private set
	var file: File? = null
		private set
	var retryPolicy: DefaultRetryPolicy? = null
		private set
	var interceptors: MutableList<Interceptor> = mutableListOf()
		private set
	var successCallbacks: MutableList<((response: VcResponse, code: String) -> Unit)> = mutableListOf()
		private set
	var errorCallbacks: MutableList<((error: VolleyError, code: String) -> Unit)> = mutableListOf()
		private set
	var cancelTag: Any? = null
		private set
	var allowLocationRedirect: Boolean = true
		private set

	init {
		setRetryPolicy(DefaultRetryPolicy(10000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT))
	}

	fun setUrl(url: String, urlEncodeSpaces: Boolean = true): InternetCall {
		this.url = if(urlEncodeSpaces){
			url.replace(" ", "%20")
		}else{
			url
		}
		return this
	}

	fun setFile(key: String, file: File): InternetCall {
		this.fileKey = key
		this.file = file
		return this
	}

	fun setRawBody(rawBody: String): InternetCall {
		this.rawBody = rawBody
		return this
	}

	fun setRawBody(json: JSONObject): InternetCall {
		putHeader("Content-type", VolleyController.ContentType.JSON.toString())
		this.rawBody = json.toString()
		return this
	}

	fun setHeaders(headers: MutableMap<String, String>): InternetCall {
		this.headers = headers
		return this
	}

	fun setParams(params: MutableMap<String, String>): InternetCall {
		rawBody = ""
		this.params = params
		return this
	}

	fun addSuccessCallback(callback: ((item: VcResponse, code: String) -> Unit)): InternetCall {
		this.successCallbacks.add(callback)
		return this
	}

	fun addErrorCallback(callback: ((error: VolleyError, code: String) -> Unit)): InternetCall {
		this.errorCallbacks.add(callback)
		return this
	}

	fun setAllowLocationRedirect(b: Boolean): InternetCall {
		allowLocationRedirect = b
		return this
	}

	fun setCode(code: String): InternetCall {
		this.code = code
		return this
	}

	fun setMethod(method: Method): InternetCall {
		this.method = method
		return this
	}

	fun replaceAccessToken(oldAccessToken: String, newAccessToken: String): InternetCall {
		url?.let { setUrl(it.replace(oldAccessToken, newAccessToken)) }

		headers.forEach { if(it.value.contains(oldAccessToken)) headers[it.key] = it.value.replace(oldAccessToken, newAccessToken) }
		params.forEach { if(it.value.contains(oldAccessToken)) headers[it.key] = it.value.replace(oldAccessToken, newAccessToken) }

		if (!rawBody.isEmpty() && rawBody.contains(oldAccessToken)) {
			rawBody = rawBody.replace(oldAccessToken, newAccessToken)
		}

		return this
	}

	/**
	 * Applies any interceptor present
	 * Should be called before build()
	 */
	fun applyInterceptors(){
		interceptors.forEach { it.intercept(this) }
	}

	fun build(listener: Response.Listener<VcResponse>, errorListener: Response.ErrorListener): Request<*> {
		val request: Request<*> = if (file == null) {
			object : CustomRequest(this.method.value(), url, errorListener) {
				override fun deliverResponse(response: Any) {
					listener.onResponse(response as VcResponse)
				}

				override fun parseNetworkResponse(response: NetworkResponse): Response<VcResponse> {
					val newCustomResponse = VcResponse(response)
					newCustomResponse.code = code
					//we set here the response (the object received by deliverResponse);
					return success(newCustomResponse, newCustomResponse.chacheHeaders)
				}

				@Throws(AuthFailureError::class)
				override fun getHeaders(): Map<String, String>? {
					return this@InternetCall.headers
				}

				override fun getParams(): Map<String, String>? {
					return this@InternetCall.params
				}

				@Throws(AuthFailureError::class)
				override fun getBody(): ByteArray {
					val body = this@InternetCall.rawBody
					if (body.isNotEmpty()) {
						return body.toByteArray()
					}
					return super.getBody() ?: "".toByteArray()
				}
			}
		} else {
			object : VolleyMultipartRequest(this.method.value(), url, errorListener) {
				override fun deliverResponse(response: Any) {
					listener.onResponse(response as VcResponse)
				}

				override fun parseNetworkResponse(response: NetworkResponse): Response<VcResponse> {
					val newCustomResponse = VcResponse(response)
					newCustomResponse.code = code
					//we set here the response (the object received by deliverResponse);
					return success(newCustomResponse, newCustomResponse.chacheHeaders)
				}

				override fun getParams(): Map<String, String>? {
					return this@InternetCall.params
				}

				override fun getByteData(): Map<String, DataPart>? {
					val byteData = mutableMapOf<String, DataPart>()
					// file name could found file base or direct access from real path
					// for now just get bitmap data from ImageView
					try {
						fileKey?.let { byteData[it] = DataPart(file?.name + "." + file?.format, ImageUtils.getFileDataFromBitmap(ImageUtils.getBitmapFromPath(file?.location)), file?.type.toString() + "/" + file?.format) }
					} catch (e: IOException) {
						e.printStackTrace()
					}

					byteData.forEach { Timber.v("${this@InternetCall.method} byteData -> ${it.key}: ${it.value}") }

					return byteData
				}
			}
		}
		request.retryPolicy = retryPolicy
		if (cancelTag != null) request.tag = cancelTag
		return request
	}

	fun setRetryPolicy(retryPolicy: DefaultRetryPolicy): InternetCall {
		this.retryPolicy = retryPolicy
		return this
	}

	fun addInterceptors(interceptors: List<Interceptor>): InternetCall {
		this.interceptors.addAll(interceptors)
		return this
	}

	fun putInterceptors(interceptors: List<Interceptor>): InternetCall {
		this.interceptors.clear()
		this.interceptors.addAll(interceptors)
		return this
	}

	/**
	 * Alias for putInterceptors
	 */
	fun setInterceptors(interceptors: List<Interceptor>): InternetCall {
		return putInterceptors(interceptors)
	}

	fun addInterceptor(interceptor: Interceptor?): InternetCall {
		if (interceptor == null) {
			return this
		}
		this.interceptors.add(interceptor)
		return this
	}

	fun putHeader(key: String, value: String): InternetCall {
		headers[key] = value
		return this
	}

	fun putParam(key: String, value: String?): InternetCall {
		if (value == null) {
			return this
		}
		rawBody = ""
		params[key] = value
		return this
	}

	fun putHeaders(headers: Map<String, String>?): InternetCall {
		if (headers == null) {
			return this
		}
		this.headers.putAll(headers)
		return this
	}

	fun putParams(params: MutableMap<String, String>): InternetCall {
		this.params.clear()
		rawBody = ""
		this.params.putAll(params)
		return this
	}

	fun addParams(params: MutableMap<String, String>): InternetCall {
		rawBody = ""
		this.params.putAll(params)
		return this
	}

	fun setCancelTag(tag: Any): InternetCall {
		this.cancelTag = tag
		return this
	}

	fun get(){
		setMethod(Method.GET)
	}

	fun post(){
		setMethod(Method.POST)
	}

	fun put(){
		setMethod(Method.PUT)
	}

	fun delete(){
		setMethod(Method.DELETE)
	}

	enum class Method {
		GET, POST, PUT, DELETE, PATCH, HEAD, OPTIONS, TRACE;

		internal fun value(): Int {
			return when (this) {
				GET -> Request.Method.GET
				POST -> Request.Method.POST
				PUT -> Request.Method.PUT
				DELETE -> Request.Method.DELETE
				PATCH -> Request.Method.PATCH
				HEAD -> Request.Method.HEAD
				OPTIONS -> Request.Method.OPTIONS
				TRACE -> Request.Method.TRACE
			}
		}

		override fun toString(): String {
			return when (this) {
				GET -> "GET"
				POST -> "POST"
				PUT -> "PUT"
				DELETE -> "DELETE"
				PATCH -> "PATCH"
				HEAD -> "HEAD"
				OPTIONS -> "OPTIONS"
				TRACE -> "TRACE"
			}
		}
	}

	fun toPostmanString(): String{
		var result = ""

		result += "Method: $method\n"
		result += "Code: $code\n"
		result += "URL: $url\n"

		if(headers.isNotEmpty()){
			result += "headers:\n"
			headers.forEach { result += "\t${it.key}: ${it.value}\n" }
		}else{
			result += "headers: none\n"
		}

		if(params.isNotEmpty()){
			result += "params:\n"
			params.forEach { result += "\t${it.key}: ${it.value}\n" }
		}else{
			result += "params: none\n"
		}

		if(rawBody.isNotEmpty()) {
			result += "body:\n"
			result += rawBody
		}else{
			result += "body: none\n"
		}

		file?.let {
			result += "file: ${file?.location}\n"
		}

		return result
	}

	interface Interceptor {
		/**
		 * @param internetCall
		 */
		fun intercept(internetCall: InternetCall)
	}
}
